game/groundforces/template.py

```diff
@@ -0,0 +1,442 @@
+
+from game.dcs.groundunittype import GroundUnitType
+from game.dcs.shipunittype import ShipUnitType
+from game.dcs.unittype import UnitType
+from game.theater.theatergroundobject import (
+    SamGroundObject,
+    EwrGroundObject,
+    BuildingGroundObject,
+    GroundGroup,
+    MissileSiteGroundObject,
+    ShipGroundObject,
+    CarrierGroundObject,
+    LhaGroundObject,
+    CoastalSiteGroundObject,
+    VehicleGroupGroundObject,
+    IadsGroundObject,
+    GroundUnit,
+)
+from game.point_with_heading import PointWithHeading
+
+from game.utils import Heading
+from game import db
+
+if TYPE_CHECKING:
+    from game import Game
+    from game.factions.faction import Faction
+    from game.theater import TheaterGroundObject, ControlPoint
+
+
+@dataclass
+class UnitTemplate:
+    name: str
+    position: Point
+    heading: int
+
+    @staticmethod
+    def from_unit(unit: Unit) -> UnitTemplate:
+        return UnitTemplate(
+            unit.name,
+            Point(int(unit.position.x), int(unit.position.y)),
+            int(unit.heading),
+        )
+
+
+@dataclass
+class GroupTemplate:
+    name: str
+    units: list[UnitTemplate]
+
+    # Is Static group
+    static: bool = False
+
+    # The group this template will be merged into
+    group: int = 1
+
+    # Define the amount of random units to be created by the randomizer.
+    # This can be a fixed int or a random value from a range of two ints as tuple
+    unit_count: list[int] = field(default_factory=list)
+
+    # defintion which unit types are supported
+    unit_types: list[str] = field(default_factory=list)
+    unit_classes: list[UnitClass] = field(default_factory=list)
+    alternative_classes: list[UnitClass] = field(default_factory=list)
+
+    # Defines if this groupTemplate is required or not
+    optional: bool = False
+
+    # Used to determine if the group should be generated or not
+    _possible_types: list[str] = field(default_factory=list)
+    _enabled: bool = True
+    _unit_type: Optional[str] = None
+    _unit_counter: Optional[int] = None
+
+    def initialize_for_faction(
+        self, faction: Faction, faction_sensitive: bool = True
+    ) -> bool:
+        # Sensitive defines if the initialization should check if the unit is available
+        # to this faction or not. It is disabled for migration only atm.
+        unit_types = [
+            t
+            for t in self.unit_types
+            if not faction_sensitive or faction.has_access_to_unit_type(t)
+        ]
+
+        alternative_types = []
+        for accessible_unit in faction.accessible_units:
+            if accessible_unit.unit_class in self.unit_classes:
+                unit_types.append(accessible_unit.dcs_id)
+            if accessible_unit.unit_class in self.alternative_classes:
+                alternative_types.append(accessible_unit.dcs_id)
+
+        if not unit_types and not alternative_types and not self.optional:
+            raise StopIteration
```

Comment:
This exception has a very specific purpose. `return False` seems appropriate based on your API?

Reply:
I've gone another way: custom Exception class `TemplateException` so that a return value is not necessary :)